libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,467 @@
+#include "AC_SafeRTL.h"
+
+/*
+*    This library is used for Copter's Safe Return-to-Launch feature. It stores
+*    "breadcrumbs" in memory, up to a certain number  of breadcrumbs have been stored,
+*    and then cleans up those breadcrumbs when space is filling up. When Safe-RTL is
+*    triggered, a more thorough cleanup occurs, and then the resulting path can
+*    is fed into WP_NAV to fly it.
+*
+*    The cleanup consists of two parts, pruning and simplification. Pruning works
+*    by calculating the closest distance reached by the line segment between any
+*    two pairs of sequential points, and cuts out anything between two points
+*    when their line segments get close. This algorithm will never compare two
+*    consecutive line segments. Obviously the segments (p1,p2) and (p2,p3) will
+*    get very close (they touch), but there would be nothing to trim between them.
+*    The simplification step uses the Ramer-Douglas-Peucker algorithm.
+*    See Wikipedia for description.
+*
+*    The simplification and pruning algorithms do not alter the path in memory,
+*    and are designed to run in the background. They each have a parameter that
+*    decides how long they will run before saving their state and returning. They
+*    are both anytime algorithms, which is helpful when memory is filling up and
+*    we just need to quickly identify a handful of points which can be deleted.
+*    The algorithms can also report if they are done. If not, the thorough cleanup
+*    will just have to wait for a bit until they are both done.
+*/
+
+SafeRTL_Path::SafeRTL_Path(bool log) :
+    accepting_new_points(true),
+    _active(true),
+    _last_index(0),
+    _simplification_complete(false),
+    _simplification_clean_until(0),
+    _pruning_complete(false),
+    _pruning_current_i(0),
+    _pruning_min_j(0),
+    _pruning_clean_until(0),
+    _logging_enabled(log)
+{
+    _simplification_bitmask = std::bitset<MAX_PATH_LEN>().set(); //initialize to 0b1111...
+    path[0] = {0.0f, 0.0f, 0.0f};
+}
+
+void SafeRTL_Path::append_if_far_enough(Vector3f p)
+{
+    if (!accepting_new_points) {
+        return;
+    }
+    if (HYPOT(p, path[_last_index]) > POSITION_DELTA) {
+        // add the breadcrumb
+        path[++_last_index] = p;
+
+        if(_logging_enabled){
```

Comment:
our formattting guidelines suggest we should have a space between the "if" and the bracket.